### ISSUE TITLE
Add GPU support to btop

### ISSUE THREAD

**Issue by Slime90** [NixOS/nixpkgs#297487]
### Describe the bug

The nixpkgs build of btop does not include GPU support as documented in https://github.com/aristocratos/btop?tab=readme-ov-file#gpu-compatibility

### Steps To Reproduce

No GPU stats shown for NVIDIA or AMD systems

### Desired behavior

GPU support for both NVIDIA and AMD

### Notify maintainers

@rmcgibbo

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by eclairevoyant** [comment]
https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/system/btop/default.nix#L10-13 you can override the package to enable support for the needed GPU(s)

**Comment by Slime90** [comment]
I missed that since it was in Master but not Unstable/Stable.

I wrote an overlay for Master and tried updating, but it did not allow me to add gpu0

I discovered that the compile flags in the docs to enable were not applied at present, so I overrode them:

```
  programs = {
    btop = {
      enable = true;
      package = pkgs.master.btop.overrideAttrs (oldAttrs: rec {
        cmakeFlags = (oldAttrs.cmakeFlags or [ ]) ++ [
          "-DBTOP_RSMI_STATIC=ON"
          "-DBTOP_GPU=ON"
        ];
      });
```

Now though, when it tries to compile, it complains that the library path isn't available

```.
building '/nix/store/lzdk9zb2vpqr1ibzhcvc9b0cbr03pm55-btop-1.3.2.drv'...
btop> Running phase: unpackPhase
btop> unpacking source archive /nix/store/q1z7rn4sgjfms2gg31nl9crqlfxpd3g2-source
btop> source root is source
btop> Running phase: patchPhase
btop> Running phase: updateAutotoolsGnuConfigScriptsPhase
btop> Running phase: configurePhase
btop> fixing cmake files...
btop> cmake flags: -DCMAKE_FIND_USE_SYSTEM_PACKAGE_REGISTRY=OFF -DCMAKE_FIND_USE_PACKAGE_REGISTRY=OFF -DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DCMAKE_INSTALL_LOCALEDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/share/locale -DCMAKE_INSTALL_LIBEXECDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/libexec -DCMAKE_INSTALL_LIBDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/lib -DCMAKE_INSTALL_DOCDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/share/doc/btop -DCMAKE_INSTALL_INFODIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/share/info -DCMAKE_INSTALL_MANDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/share/man -DCMAKE_INSTALL_OLDINCLUDEDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/include -DCMAKE_INSTALL_INCLUDEDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/include -DCMAKE_INSTALL_SBINDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/sbin -DCMAKE_INSTALL_BINDIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/bin -DCMAKE_INSTALL_NAME_DIR=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2/lib -DCMAKE_POLICY_DEFAULT_CMP0025=NEW -DCMAKE_OSX_SYSROOT= -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_STRIP=/nix/store/qhpw32pz39y6i30b3vrbw5fw6zv5549f-gcc-wrapper-13.2.0/bin/strip -DCMAKE_RANLIB=/nix/store/qhpw32pz39y6i30b3vrbw5fw6zv5549f-gcc-wrapper-13.2.0/bin/ranlib-DCMAKE_AR=/nix/store/qhpw32pz39y6i30b3vrbw5fw6zv5549f-gcc-wrapper-13.2.0/bin/ar -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++ -DCMAKE_INSTALL_PREFIX=/nix/store/yq0fpm8503mcl6i8yk7vqlv01wy6jsmb-btop-1.3.2 -DBTOP_RSMI_STATIC=ON -DBTOP_GPU=ON
btop> -- The CXX compiler identification is GNU 13.2.0
btop> -- Detecting CXX compiler ABI info
btop> -- Detecting CXX compiler ABI info - done
btop> -- Check for working CXX compiler: /nix/store/qhpw32pz39y6i30b3vrbw5fw6zv5549f-gcc-wrapper-13.2.0/bin/g++ - skipped
btop> -- Detecting CXX compile features
btop> -- Detecting CXX compile features - done
btop> -- Looking for C++ include ranges
btop> -- Looking for C++ include ranges - found
btop> -- Performing Test HAS_FSTACK_PROTECTOR
btop> -- Performing Test HAS_FSTACK_PROTECTOR - Success
btop> -- Performing Test HAS_FCF_PROTECTION
btop> -- Performing Test HAS_FCF_PROTECTION - Success
btop> -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
btop> -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
btop> -- Found Threads: TRUE
btop> CMake Error at CMakeLists.txt:145 (add_subdirectory):
btop>   add_subdirectory given source "lib/rocm_smi_lib" which is not an existing
btop>   directory.
btop>
btop> -- Configuring incomplete, errors occurred!
error: builder for '/nix/store/lzdk9zb2vpqr1ibzhcvc9b0cbr03pm55-btop-1.3.2.drv' failed with exit code 1
```

Correct me if I am wrong, but doesn't this conditional  to include the library path occurring too late in the stages?

```
  postPhases = lib.optionals rocmSupport [ "postPatchelf" ];
  postPatchelf = lib.optionalString rocmSupport ''
    patchelf --add-rpath ${lib.getLib rocmPackages.rocm-smi}/lib $out/bin/btop
  '';
```

This path is needed before patchelf by the compiler......

**Comment by SomeoneSerge** [comment]
> `-DBTOP_RSMI_STATIC=<ON\|OFF>`  | Build and link the ROCm SMI library statically (OFF by default) 

1. https://github.com/NixOS/nixpkgs/pull/296468 uses the shared library instead; does that work for you?
1. Upstream is broken/uses README instead of `find_package` to locate `rocm_smi_lib`: https://github.com/aristocratos/btop/blob/c767099d765b0094c50b8c66030aeacff26f56ef/README.md?plain=1#L383-L387

**Comment by Slime90** [comment]
>     1. [btop: enable ROCm support #296468](https://github.com/NixOS/nixpkgs/pull/296468) uses the shared library instead; does that work for you?
> 

No, it did not work for me on my ryzen 6000 series laptop APU w/ integrated zen amdgpu. Its launches and the GPU option is in the menu, but it errors when trying to add gpu0 to the layout. Others had mentioned in issues submitted to the btop repo that this could happen when the shared/dynamic library was not present or readable by btop. This is why I tried the static linked option.

I can try the above test again on a system with a dedicated firepro card, in case this is an issue specifically with the zen APU GPU.

**Comment by mguentner** [comment]
@Slime90 Do have `rocmSupport` set to `true` in your nixpkgs config?

https://nixos.org/manual/nixpkgs/unstable/#opt-rocmSupport

**Comment by mguentner** [comment]
> Correct me if I am wrong, but doesn't this conditional to include the library path occurring too late in the stages?

It's a bit weird tbh. [btop defines everything itself if you do not do static linking](https://github.com/aristocratos/btop/blob/main/src/linux/btop_collect.cpp#L160) which is not done by the derivation.

It is then doing the dynamic linking itself: https://github.com/aristocratos/btop/blob/main/src/linux/btop_collect.cpp#L1283 using `dlopen`.

**Comment by Slime90** [comment]
> @Slime90 Do have `rocmSupport` set to `true` in your nixpkgs config?
> 
> https://nixos.org/manual/nixpkgs/unstable/#opt-rocmSupport

Yes, I do. I know that the option had an effect because without it the GPU options are not select-able in the btop settings menu, and with it set to true I get the config option.

Now the issue seems to be that btop cannot utilize the library properly, since I get an error for "invalid box names" exactly as reported in https://github.com/aristocratos/btop/issues/784

**Comment by LDprg** [comment]
I would be nice if btop had some option to enable and disable gpu support.

**Comment by Slime90** [comment]
> It would be nice if btop had some option to enable and disable gpu support.

This (as I found out in this thread) is provided by the nixpkgs option discussed above. This is certainly satisfactory since this is a package.

The issue is that at least in my testing, the option doesn't fully work.

**Comment by mguentner** [comment]
Can you try compiling btop and linking statically? You might also try using btop's `Makefile` instead of directly using `cmake` like the derivation does.

**Comment by Slime90** [comment]
Since I didn't post it previously, here is the error from the debug log on startup, with `nixpkgs.config.rcomSupport = true`:

`2024/03/24 (14:44:02) | INFO: Failed to load librocm_smi64.so, AMD GPUs will not be detected: librocm_smi64.so.6: cannot open shared object file: No such file or directory`

**Comment by Slime90** [comment]
Looks like this might be fixed by #298135 which was opened 2 days ago

**Comment by Slime90** [comment]
I made 2 errors that led to this not working.

1. I made a typo in the nixpkgs config option attribute name, `rcom` instead of `rocm`, which led to the option not being evaluated. It makes sense to me now why this fails silently and does not error, but at the time I was not considering that possibility

2. As many others do, I am running nixos stable, with an overlay for selectively including unstable packages. I had forgotten (as I occasionally do) that nixpkgs options which are set do not apply to overlays for other versions of nixpkgs when configured in this specific way. 

I believe this issue can be closed, both unstable and master versions of the package are working w/ the option properly set in my unstable overlay. Sorry for the confusion and hopefully this helps someone else.
